lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -993,22 +1001,98 @@ class MongoClientInterface {
         return undefined;
     }
 
+    /*
+     * Execute the user-defined query in a stage then sort it for
+     * stateless paging. The output is stored in a temporary
+     * collection in a special namespace that will be periodically
+     * erased (e.g. once a day).
+     *
+     * All search queries are bounded by MAX_TIME_MS env (default is
+     * 5mn).
+     */
+    doSearch(c, tempCollection, params, extension, searchOptions, log, cb) {
+        const _cursor = c.aggregate([
+            { $match: searchOptions }, // user query
+            { $sort: { _id: 1 } }, // needed for paging
+            { $out: tempCollection }, // a job will clean it up
+        ],
+            {
+                maxTimeMs: this.maxTimeMs, // always limit
+                allowDiskUse: true, // stage large queries on disk
+            },
+            null);
+        _cursor.toArray(err => {
```

Comment:
No the result is empty because there is a `$out`.

Reply:
But one problem of this approach is that it returns nothing until the query is done. This is weird from the API client perspective. Any suggestion ?

Reply:
If it's streaming the output, may be use a combined approach of streaming 1000 documents back to the client and starting a background async job that writes to the temporary collection
mongodb.github.io/node-mongodb-native/3.6/reference/cursors/#stream-api
It sounds complex and dirty, ideally I would delegate the task to some other worker but I can't see how.

Reply:
Yes we could to basically the user-query in an aggregate AND in a find() at the same time. But it will execute the query twice... just to avoid sessions, that's a bit too much...

Reply:
Actually the $sort here is also killing the query plan, and the good news is ... no need to do the $sort because in the $out collection the keys will already be sorted...

Reply:
Need to also limit it.